internal/app.go

```diff
@@ -0,0 +1,187 @@
+		return
+	}
+	resp, err := http.Get(fmt.Sprintf(src, stud, token))
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	all, err := io.ReadAll(resp.Body)
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	cal, err := ics.ParseCalendar(strings.NewReader(string(all)))
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
+	hasLecture := make(map[string]bool)
+	var newComponents []ics.Component
+	for _, component := range cal.Components {
+		switch component.(type) {
+		case *ics.VEvent:
+			event := component.(*ics.VEvent)
+			dedupKey := fmt.Sprintf("%s-%s", event.GetProperty(ics.ComponentPropertySummary).Value, event.GetProperty(ics.ComponentPropertyDtStart))
+			if _, ok := hasLecture[dedupKey]; ok {
+				log.Println("skipping ", dedupKey)
+				continue
+			}
+			hasLecture[dedupKey] = true
+			a.cleanEvent(event)
+			newComponents = append(newComponents, event)
+		default:
+			newComponents = append(newComponents, component)
+		}
+	}
+	cal.Components = newComponents
+
+	response := []byte(cal.Serialize())
+	c.Header("Content-Type", "text/calendar")
+	c.Header("Content-Length", fmt.Sprintf("%d", len(response)))
+	c.Writer.Write(response)
+}
+
+// matches tags like (IN0001) or [MA2012] and everything after.
+// unfortunate also matches wrong brackets like [MA123) but hey…
+var reTag = regexp.MustCompile(" ?[\\[(](MA|IN|WI|WIB)[0-9]+((_|-|,)[a-zA-Z0-9]+)*[\\])].*")
+
+// Matches location and teacher from language course title
+var reLoc = regexp.MustCompile(" ?(München|Garching|Weihenstephan).+")
+
+// Matches repeated whitespaces
+var reSpace = regexp.MustCompile(`\s\s+`)
+
+var unneeded = []string{
+	"Standardgruppe",
+	"PR",
+	"VO",
+	"FA",
+	"VI",
+	"TT",
+	"UE",
+	"SE",
+	"(Limited places)",
+	"(Online)",
+}
+
+var reRoom = regexp.MustCompile("^(.*?),.*(\\d{4})\\.(?:\\d\\d|EG|UG|DG|Z\\d|U\\d)\\.\\d+")
+
+func (a *App) cleanEvent(event *ics.VEvent) {
+	summary := strings.ReplaceAll(event.GetProperty(ics.ComponentPropertySummary).Value, "\\", "")
+	description := strings.ReplaceAll(event.GetProperty(ics.ComponentPropertyDescription).Value, "\\", "")
+	location := strings.ReplaceAll(event.GetProperty(ics.ComponentPropertyLocation).Value, "\\", "")
+
+	//Remove the TAG and anything after e.g.: (IN0001) or [MA0001]
+	summary = reTag.ReplaceAllString(summary, "")
+	//remove location and teacher from language course title
+	summary = reLoc.ReplaceAllString(summary, "")
+	summary = reSpace.ReplaceAllString(summary, "")
+	for _, replace := range unneeded {
+		summary = strings.ReplaceAll(summary, replace, "")
+	}
+
+	/*
+			todo, whatever this does:
+			//Clean up extra info for language course names
+		    if(preg_match('/(Spanisch|Französisch)\s(A|B|C)(1|2)((\.(1|2))|(\/(A|B|C)(1|2)))?(\s)/', $summary, $matches, PREG_OFFSET_CAPTURE) === 1){
+		       $summary = substr($summary, 0, $matches[10][1]);
+		    }
```

Comment:
So, this matches on
```
Französisch B2 
Spanisch A1.1 
Spanisch A1/A2 
Spanisch A1 
```

And then shortens `$summary` to the tenth capture group's length.
The tenth capture group is `(\s)`… So, basically, it shortens it to :
```
F
S
S
S
```

Reply:
But why

Reply:
Lets leave this out for now.